Feature/basic calculation

dhruv-d-simform opened the pull request:
# Description

- Implemented the basic functionality of Calculator for these operations: `+ - * /`
- Added click and keyboard event handlers to take input for Calculator.
- Used `eval` function to evaluate expressions.
- Implemented a `Worker` that executes `eval` function to isolate the execution of `eval` function.
- Maintained the modularity throughout the code.

# Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Documentation update
- [ ] Refractoring
- [ ] Performance improvement
- [ ] Scaffolding

# Screenshots

![image](https://github.com/user-attachments/assets/3fc3b935-d933-4bbf-b84c-8939073fba3d)

# Validations

![Screenshot from 2025-03-11 16-12-07](https://github.com/user-attachments/assets/33dd1809-ce9f-4552-b179-3c3c23993447)

![Screenshot from 2025-03-11 16-12-56](https://github.com/user-attachments/assets/08278196-6d64-4641-a41d-8cdbf1743cf4)

![Screenshot from 2025-03-11 16-13-36](https://github.com/user-attachments/assets/3a5464db-b504-4ad0-b08f-1b99c2abf7f3)

# Technical Context:

### Usage of `Worker` to implement expression evaluation.  
Used it to achieve 3 things:
1. To isolate the execution of `eval` function, which will be good for security. Because `Worker` will not have access to DOM or localStorage or Cookies.
2. To maintain modularity of code.
3. To be able to create global variables like `sin`, `cos` to use them directly inside `eval` without having to use `Math.sin`. We will not pollute global environment of regular script by using `Worker`.

# How to test

Checkout the deployment preview: <https://deploy-preview-3--scific-calc.netlify.app/>

dhruv-d-simform committed:
feature: Create Display class to handle calculator's display

dhruv-d-simform committed:
feature: Create script for evaluation of expressions

dhruv-d-simform committed:
feature: Create Calculator functionality with basic operations :- + - * /

dhruv-d-simform committed:
feature: Handle parenthesis in Calculator

dhruv-d-simform committed:
feature: Add keyboard events to give input to calculator

dhruv-d-simform committed:
fix: Remove the defer attribute from script tag.

dhruv-d-simform committed:
feature: Show '0' on display when clearing it

dhruv-d-simform committed:
fix: Show '0' when display is clear

dhruv-d-simform committed:
fix: Throw error on divide by '0'

dhruv-d-simform committed:
refractor: Create a seperate function to handle input from both keyboard and button click in Calculator class

dhruv-d-simform committed:
fix: Add border radius to body and buttons

dhruv-d-simform committed:
fix: Remove extra spaces

dhruv-d-simform committed:
feature: Add character limit for input

Kaushal-SF commented:
@dhruvdhaduk-simform Observed few things in the UI.

1) Maybe we can improve styling a bit for this. Just curious, why we didn't use similar symbol available in mock-up?

<img width="129" alt="image" src="https://github.com/user-attachments/assets/472d1891-d40f-4305-aed1-c46b45987ae5" />

2) When we reach to maximum input limit, is it good practice to clear the input, when we close the pop-up? Thoughts?

<img width="536" alt="image" src="https://github.com/user-attachments/assets/580e7ea8-05d2-45e3-9f41-ea6a136c9472" />

In addition to that, when I observed another anomaly. Let's say I have added 24 numbers and then adding it with a 5 number, it is giving me an error of maximum limit reached. I think this shouldn't be the case.
ex: 111111111111111111111111 + 33 (Will get the pop-up for max limit reached)

Kaushal-SF commented on scripts/workers/evaluator.js:
```diff
@@ -0,0 +1,36 @@
+        // Handle divide by 0
+        if (Math.abs(result) === Infinity) {
+            throw new Error("This operation is not allowed");
+        }
```
Can we use the `isFinite` method instead of checking in this `Math.abs(result) === Infinity`?
```suggestion
        if (!isFinite(result)) {
            throw new Error("This operation is not allowed");
        }
```

dhruv-d-simform replied:
Used the `isFinite` function. 
https://github.com/dhruvdhaduk-simform/scientific-calc/pull/3/commits/df6736e08c3a1b05886a175813c7f6a17458da7f

Kaushal-SF commented on scripts/display.js:
```diff
@@ -0,0 +1,52 @@
+"use strict";
+
+const CHARACTER_LIMIT = 25;
```
Good usage of `screaming snake case` for constant 👍

Kaushal-SF commented on scripts/index.js:
```diff
@@ -0,0 +1,6 @@
+import { Calculator } from "./calculator.js";
+
+// Initialize the calculator.
+let calc = new Calculator("display-text", "control");
```
We can be specific about the variable name.

```suggestion
let calculator = new Calculator("display-text", "control");
```

dhruv-d-simform replied:
Fixed it.
https://github.com/dhruvdhaduk-simform/scientific-calc/pull/3/commits/169d2f47b90ab568975b7e9b35d91ae50c3b4787

Kaushal-SF commented on scripts/calculator.js:
```diff
@@ -0,0 +1,90 @@
+        this.btns = document.querySelector(`#${btnsID}`);
+        this.abortController = new AbortController();
+
+        this.evaluator = new Worker("/scripts/workers/evaluator.js");
```
Good usage of `Worker` to move expensive calculations off the main thread 👍

Kaushal-SF commented on scripts/workers/evaluator.js:
```diff
@@ -0,0 +1,36 @@
+        globalThis.postMessage({
+            success: false,
+            q: event.data,
+            error
```
Can we convert error to string to ensure that it always be available in string format.

```suggestion
            error: error.toString()
```

dhruv-d-simform replied:
Fixed it.
https://github.com/dhruvdhaduk-simform/scientific-calc/pull/3/commits/d9fe68a6dd9b2987e47c6c845269c90fd685995f

Kaushal-SF commented on scripts/calculator.js:
```diff
@@ -0,0 +1,90 @@
+
+    // Remove all event handlers and terminate evaluator worker.
+    destroy() {
+        this.abortController?.abort();
+        this.evaluator.terminate();
```
Can we also check that `this.evaluator` is defined before terminating it?

dhruv-d-simform replied:
Added check before terminating evaluator.
https://github.com/dhruvdhaduk-simform/scientific-calc/pull/3/commits/a4fd37f03d809e45e8bca1d5789ec69db2335876

Kaushal-SF commented:
Can we also avoid multiple decimal points in the input?

<img width="475" alt="image" src="https://github.com/user-attachments/assets/9f40d32a-8497-43f2-9656-8819f93ada9a" />

dhruv-d-simform committed:
fix: Add character limit to numbers instead of whole input

dhruv-d-simform committed:
refractor: Use 'isFinite' function instead of checking the result against 'Infinity'

dhruv-d-simform committed:
refractor: Use the full name 'calculator' for variable 'calc' in index.js

dhruv-d-simform committed:
fix: Convert the error to string before passing it to main script from evaluator worker

dhruv-d-simform committed:
fix: Check if the evaluator is defined before terminating it in Calculator class

dhruv-d-simform committed:
fix: Prevent user from entering a continuous sequence of 'dots'

dhruv-d-simform committed:
docs: Add few comments regarding validation in Display class

dhruv-d-simform committed:
fix: Change the symbol for 'plus or minus' in index.html

dhruv-d-simform commented:
> 1. Maybe we can improve styling a bit for this. Just curious, why we didn't use similar symbol available in mock-up?

Fixed it. https://github.com/dhruvdhaduk-simform/scientific-calc/pull/3/commits/32fbf933af10369eac75b79ee684cb6c671dc51a

dhruv-d-simform commented:
> 2. When we reach to maximum input limit, is it good practice to clear the input, when we close the pop-up? Thoughts?
> 
> <img alt="image" width="536" src="https://private-user-images.githubusercontent.com/97082607/424009801-580e7ea8-05d2-45e3-9f41-ea6a136c9472.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIzNjQ1OTEsIm5iZiI6MTc0MjM2NDI5MSwicGF0aCI6Ii85NzA4MjYwNy80MjQwMDk4MDEtNTgwZTdlYTgtMDVkMi00NWUzLTlmNDEtZWE2YTEzNmM5NDcyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE5VDA2MDQ1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVjY2RhMmYwNzYzODY4NWMxOTI0YTZjNjA0OWUwN2M0MTU3ZDIwM2JhZGI4Yjc1YjM5ZGNhY2U1ODlhMjA5YjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.FV5lH-ehDUPRhUdq0VnsTibP6syYWqHrg2gRRrAmB0w">
> In addition to that, when I observed another anomaly. Let's say I have added 24 numbers and then adding it with a 5 number, it is giving me an error of maximum limit reached. I think this shouldn't be the case. ex: 111111111111111111111111 + 33 (Will get the pop-up for max limit reached)

I had added limit to whole input. So that's why this issue occured. I have fixed it by add character limit to numbers only. (This makes sense because when we will have functions like sin, cos, round, floor, they will consume most of the limit.)
https://github.com/dhruvdhaduk-simform/scientific-calc/pull/3/commits/3608902990e0fa7059d2b99ecfc71aa63f8a2b70

dhruv-d-simform commented:
> Can we also avoid multiple decimal points in the input?
> 
> <img alt="image" width="475" src="https://private-user-images.githubusercontent.com/97082607/424296071-9f40d32a-8497-43f2-9656-8819f93ada9a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIzNjQ1OTEsIm5iZiI6MTc0MjM2NDI5MSwicGF0aCI6Ii85NzA4MjYwNy80MjQyOTYwNzEtOWY0MGQzMmEtODQ5Ny00M2YyLTk2NTYtODgxOWY5M2FkYTlhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE5VDA2MDQ1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ2OWRiYmVmM2Q0MjQ1OWUyMGQwZTdlMTdjNTE4NDljMjkwZTA2ZGE0NjkyNDk3NGI5ZDQwZTY3NzU3OTBlNzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ZRhLHGcydvPXwnyoF1pV0kshB4edGiLrhKjvrs8Cgvg">

Fixed it.
https://github.com/dhruvdhaduk-simform/scientific-calc/pull/3/commits/34b1701713dc2a41041ee8dc43de42ac51891d83

dhruv-d-simform committed:
fix: Prevent user from entering multiple operators

dhruv-d-simform committed:
fix: Prevented user from entering a number with multiple dots

Kaushal-SF commented on scripts/calculator.js:
```diff
@@ -0,0 +1,90 @@
+            case "-":
+            case "*":
+            case "/":
+            case ".":
```
Can we handle the multiple decimal input issue directly with `regex`? Replacing [this](https://github.com/dhruvdhaduk-simform/scientific-calc/pull/3/commits/34b1701713dc2a41041ee8dc43de42ac51891d83#diff-ad1f93112bef47b879633d364b2363c5a1c0e6e298a22697394629b903f38301R37-R51) functionality entirely.

```suggestion
            case ".":
                if (this.display.get().match(/\.\d*$/)) return;
                this.display.append(input);
                return;
```

dhruv-d-simform replied:
Used the regex. However, I placed this validation inside `append` function of `display.js` where all validations are implemented.
https://github.com/dhruvdhaduk-simform/scientific-calc/pull/3/commits/aff247030e3c715d6f45d7fe06b4686b986c531f

Kaushal-SF commented on scripts/display.js:
```diff
@@ -0,0 +1,118 @@
+        }
+
+        // Prevent user from entering multiple operators.
+        const operators = ['+', '-', '*', '/'];
+        const newInput = this.get() + txt;
+        let isOperator = false;
+        for (let i = newInput.length - 1; i >= 0; i--) {
+            if (operators.includes(newInput[i])) {
+                if (isOperator) {
+                    return;
+                }
+                isOperator = true;
+            }
+            else {
+                isOperator = false;
+            }
+        }
```
This logic looks good. Could we utilize the power of `Set` here? To make the code more robust.
Leaving it up to you if you want to give it a try.

dhruv-d-simform replied:
Used `Set` to hold operators.
https://github.com/dhruvdhaduk-simform/scientific-calc/pull/3/commits/f37b06983202761d9a93cd4c00f7f8fa36293bed

Kaushal-SF commented on scripts/display.js:
```diff
@@ -0,0 +1,118 @@
+
+        // Prevent multiple dots inside a number.
+        let isInsideNumber = false;
+        let isDotOccured = false;
```
Naming suggestion:
```suggestion
        let inNumberSequence = false;
        let hasDecimal = false;
```

dhruv-d-simform replied:
The regex implemented in this commit(https://github.com/dhruvdhaduk-simform/scientific-calc/pull/3/commits/aff247030e3c715d6f45d7fe06b4686b986c531f) handles this validation. So I removed this whole block.

Kaushal-SF reviewed (commented):
Added few final comments.

dhruv-d-simform committed:
refractor: Use regex to validate user input regarding multiple decimal points

dhruv-d-simform committed:
refractor: Use Set to hold operators in validation of user input

Kaushal-SF reviewed (approved):
Changes looks good 👍
